File: ts/src/rrule/expand.ts
// SPDX-License-Identifier: MIT

// The four expansion entry points — `nextOccurrence`, `all`,
// `occurrences`, `between` — plus the shared walk behind them.
//
// Spec §Expansion names three bounding strategies and this module
// offers all three: a count (`occurrences`), a window (`between`), and
// laziness (`all`, which the consumer stops). An unbounded request
// through any of the bounded two is `ErrUnboundedExpansion`, not an
// empty result.

import type { Instant } from "../date.js";
import { VstarError } from "../errors.js";
import type { Expansion, RuleFields } from "./types.js";
import { MAX_ITERATIONS, advance, periodOccurrences } from "./evaluate.js";

export { MAX_ITERATIONS };

/**
 * Reject a rule the evaluator cannot walk.
 *
 * `parseRRule` guarantees both conditions, so this only fires for a
 * rule literal assembled by hand — which is exactly why it reports
 * `ErrUnsupportedRRule` rather than trusting the value.
 */
export function checkExpandable(rule: RuleFields): void {
  if (rule.freq === "INVALID") {
    throw new VstarError("ErrUnsupportedRRule", "rrule: FREQ is required");
  }
  if (rule.interval < 1) {
    throw new VstarError("ErrUnsupportedRRule", `rrule: INTERVAL must be >= 1, got ${rule.interval}`);
  }
}

/** The `ErrIterationCap` failure, naming the bound that was hit. */
export function iterationCap(op: string, rule: RuleFields): VstarError {
  return new VstarError(
    "ErrIterationCap",
    `rrule: ${op}: no occurrence within ${MAX_ITERATIONS} consecutive ${rule.freq} periods`,
  );
}

/** How a {@link walk} ended. */
export interface WalkResult {
  /**
   * `true` when {@link MAX_ITERATIONS} consecutive periods produced
   * nothing and the search was abandoned. This is NOT termination — the
   * rule may well have occurrences beyond the budget.
   */
  readonly capped: boolean;
}

/**
 * Yield every occurrence of `rule` from `dtstart`, chronologically,
 * into `yield_`, stopping when it returns `false`.
 *
 * The iteration bound is a starvation guard: the empty-period counter
 * resets whenever a period yields, so a regularly-firing rule runs as
 * long as the caller wants while one whose BY-* clauses can never match
 * still stops.
 *
 * An unwalkable rule yields nothing and is not a cap hit; callers that
 * need that reported run {@link checkExpandable} first.
 */
export function walk(rule: RuleFields, dtstart: Instant, yield_: (t: Instant) => boolean): WalkResult {
  try {
    checkExpandable(rule);
  } catch {
    return { capped: false };
  }

  let emitted = 0;
  let empty = 0;
  let current: Instant | undefined = dtstart;

  while (empty < MAX_ITERATIONS) {
    let produced = false;
    for (const occ of periodOccurrences(rule, current, dtstart)) {
      // A period can reach back before dtstart — a weekly expansion
      // covers the whole week dtstart falls in — and the series starts
      // at dtstart, never earlier.
      if (occ < dtstart) continue;
      // UNTIL is inclusive per RFC 5545 §3.3.10.
      if (rule.until !== undefined && occ > rule.until) return { capped: false };
      produced = true;
      emitted++;
      if (!yield_(occ)) return { capped: false };
      if (rule.count > 0 && emitted >= rule.count) return { capped: false };
    }
    empty = produced ? 0 : empty + 1;
    current = advance(rule, current);
    if (current === undefined) return { capped: false };
  }
  return { capped: true };
}

/**
 * The next occurrence strictly after `after`, or `undefined` when the
 * rule has terminated.
 *
 * Termination — UNTIL passed, COUNT exhausted, no further period — is
 * `undefined` and is normal completion. Reaching the iteration bound is
 * `ErrIterationCap` and is not: the rule has not necessarily ended, the
 * evaluator stopped looking. A caller that treats the two alike
 * silently drops occurrences, which is why they are different answers.
 *
 * The first occurrence of a rule is `dtstart` itself whenever it
 * satisfies the BY-* filters, per RFC 5545 — pass `after = dtstart` to
 * step past it.
 */
export function nextOccurrence(rule: RuleFields, dtstart: Instant, after: Instant): Instant | undefined {
  checkExpandable(rule);

  let found: Instant | undefined;
  const result = walk(rule, dtstart, (occ) => {
    if (occ > after) {
      found = occ;
      return false;
    }
    return true;
  });
  if (found !== undefined) return found;
  if (result.capped) throw iterationCap("nextOccurrence", rule);
  return undefined;
}

/**
 * Every occurrence of `rule` from `dtstart`, lazily.
 *
 * A rule with neither UNTIL nor COUNT is infinite, and iterating this
 * without breaking will not return — that is the documented behaviour
 * of a lazy sequence and the reason it exists. Use {@link occurrences}
 * or {@link between} when a bounded result is what you want.
 *
 * The sequence also ends at the iteration bound. A generator has
 * nowhere to put an error, so `all` cannot tell that apart from
 * termination; the bounded entry points report it as
 * `ErrIterationCap`.
 */
export function* all(rule: RuleFields, dtstart: Instant): Generator<Instant, void, undefined> {
  // The walk is push-shaped and a generator is pull-shaped, so the
  // occurrences of one period are buffered and drained rather than
  // re-entering the walk per value.
  try {
    checkExpandable(rule);
  } catch {
    return;
  }

  let emitted = 0;
  let empty = 0;
  let current: Instant | undefined = dtstart;

  while (empty < MAX_ITERATIONS) {
    let produced = false;
    for (const occ of periodOccurrences(rule, current, dtstart)) {
      if (occ < dtstart) continue;
      if (rule.until !== undefined && occ > rule.until) return;
      produced = true;
      emitted++;
      yield occ;
      if (rule.count > 0 && emitted >= rule.count) return;
    }
    empty = produced ? 0 : empty + 1;
    current = advance(rule, current);
    if (current === undefined) return;
  }
}

/**
 * Up to `limit` occurrences, with the flag that says which way the
 * expansion stopped.
 *
 * `complete` is `true` when the rule itself terminated within the limit
 * — the returned list is the entire series — and `false` when the limit
 * truncated it. Reporting that distinction is required by spec
 * §Expansion: a caller that stops after N steps otherwise never learns
 * whether N was the whole series or merely the first N.
 *
 * A limit of `0` returns no occurrences and `complete: false` — no
 * occurrences, and no claim that the series ended. A negative limit is
 * `ErrUnboundedExpansion`.
 */
export function occurrences(rule: RuleFields, dtstart: Instant, limit: number): Expansion {
  if (limit < 0) {
    throw new VstarError(
      "ErrUnboundedExpansion",
      `rrule: occurrences: limit must be >= 0, got ${limit}`,
    );
  }
  checkExpandable(rule);
  if (limit === 0) return { times: [], complete: false };

  const out: Instant[] = [];
  let truncated = false;
  const result = walk(rule, dtstart, (occ) => {
    if (out.length === limit) {
      // The walk produced one more than asked for, so the series
      // definitively continues past the limit.
      truncated = true;
      return false;
    }
    out.push(occ);
    return true;
  });
  if (result.capped) throw iterationCap("occurrences", rule);
  return { times: out, complete: !truncated };
}

/**
 * Every occurrence in the half-open window `[start, end)` — `start`
 * inclusive, `end` exclusive.
 *
 * The window bounds the result, so this terminates even for a rule with
 * neither UNTIL nor COUNT. It does not bound the *search*: a rule that
 * never yields never reaches `end`, so the iteration bound stops it and
 * the failure is `ErrIterationCap`.
 *
 * A missing `end`, or an `end` not strictly after `start`, is
 * `ErrUnboundedExpansion` rather than a silent empty result — an
 * unbounded window is an infinite expansion request.
 */
export function between(rule: RuleFields, dtstart: Instant, start: Instant, end: Instant): Instant[] {
  if (end === undefined || end === null || Number.isNaN(end)) {
    throw new VstarError(
      "ErrUnboundedExpansion",
      "rrule: between: end must be present (an open-ended window is an infinite expansion; use all)",
    );
  }
  if (!(end > start)) {
    throw new VstarError("ErrUnboundedExpansion", `rrule: between: end ${end} must be after start ${start}`);
  }
  checkExpandable(rule);

  const out: Instant[] = [];
  const result = walk(rule, dtstart, (occ) => {
    if (occ >= end) return false;
    if (occ >= start) out.push(occ);
    return true;
  });
  if (result.capped) throw iterationCap("between", rule);
  return out;
}

File: go/rrule/expand.go
// SPDX-License-Identifier: MIT

package rrule

import (
	"errors"
	"fmt"
	"iter"
	"slices"
	"time"
)

// ErrUnboundedExpansion signals an expansion request that has no
// terminating condition: neither the rule (UNTIL/COUNT), nor a
// caller-supplied limit, nor a finite window bounds the result.
//
// Expansion into a slice MUST be bounded — a FREQ=DAILY rule with
// no UNTIL and no COUNT has infinitely many occurrences. Callers
// that genuinely want an open-ended series use All, which is lazy.
//
// Consumers match with errors.Is.
var ErrUnboundedExpansion = errors.New("rrule: unbounded expansion request")

// All returns a lazy iterator over every occurrence of rule
// starting at dtstart, in chronological order.
//
// Termination contract:
//
//   - The sequence ends on its own when the rule terminates (UNTIL
//     reached — inclusive per RFC 5545 — or COUNT exhausted).
//   - A rule with neither UNTIL nor COUNT is INFINITE. Ranging
//     over All without breaking will not return. This is the
//     documented, intended behavior of a lazy iterator; use
//     Occurrences or Between when you need a bounded slice.
//   - The sequence also ends when the evaluator's iteration bound
//     is reached: MaxIterations consecutive FREQ periods without an
//     occurrence (an unsatisfiable BY-* combination such as
//     FREQ=YEARLY;BYMONTH=2;BYMONTHDAY=30). A range-over-func
//     iterator has nowhere to put an error, so All cannot tell
//     this apart from termination; Occurrences and Between report
//     it as ErrIterationCap.
//   - An invalid rule (FREQ unset, INTERVAL < 1) yields nothing.
//     Use Occurrences or Between if you need that reported as an
//     error.
//
// The first occurrence is dtstart itself whenever dtstart
// satisfies the rule's BY-* filters, per RFC 5545 §3.3.10.
//
// Time-zone semantics match NextOccurrence: candidates carry
// dtstart's location, so a zoned dtstart preserves wall-clock time
// across DST transitions and the UTC offset shifts accordingly.
func All(rule Rule, dtstart time.Time) iter.Seq[time.Time] {
	return func(yield func(time.Time) bool) {
		walk(rule, dtstart, yield)
	}
}

// Occurrences returns up to limit occurrences of rule starting at
// dtstart, in chronological order.
//
// The complete return value distinguishes the two ways expansion
// stops:
//
//   - complete == true: the rule itself terminated (UNTIL reached
//     or COUNT exhausted) within the limit, so the returned slice
//     is the ENTIRE series. Nothing was truncated.
//   - complete == false: the limit was reached first, so more
//     occurrences exist beyond the last element returned.
//
// This is the distinguishable termination signal that a bare
// stepping loop cannot provide: a caller that stops after N steps
// never learns whether N was the whole series or just the first N.
//
// limit must be >= 0. A limit of 0 returns (nil, false, nil) — no
// occurrences, and no claim that the series ended.
//
// Errors:
//
//   - wrapped ErrIterationCap when the evaluator walked
//     MaxIterations consecutive FREQ periods without an occurrence
//     and abandoned the search. The rule has not terminated; the
//     evaluator gave up, and complete is false.
//   - wrapped ErrUnboundedExpansion for a negative limit.
//   - wrapped ErrUnsupportedRRule when rule is not evaluable (FREQ
//     unset or INTERVAL < 1), which happens only for Rule literals
//     that did not go through ParseRRule.
func Occurrences(rule Rule, dtstart time.Time, limit int) (occs []time.Time, complete bool, err error) {
	if limit < 0 {
		return nil, false, fmt.Errorf("rrule: Occurrences: limit must be >= 0, got %d: %w", limit, ErrUnboundedExpansion)
	}
	if err := checkExpandable(rule); err != nil {
		return nil, false, err
	}
	if limit == 0 {
		return nil, false, nil
	}

	out := make([]time.Time, 0, limit)
	truncated := false
	capped := walk(rule, dtstart, func(occ time.Time) bool {
		if len(out) == limit {
			// The walk produced one more than we asked for, so the
			// series definitively continues past the limit.
			truncated = true
			return false
		}
		out = append(out, occ)
		return true
	})
	if capped {
		return nil, false, iterationCapError("Occurrences", rule)
	}
	if len(out) == 0 {
		return nil, !truncated, nil
	}
	return out, !truncated, nil
}

// Between returns every occurrence of rule in the half-open window
// [start, end) — start inclusive, end exclusive — in chronological
// order.
//
// Termination contract: the window itself bounds the result, so
// Between always terminates even for a rule with no UNTIL and no
// COUNT. Expansion stops at the earlier of `end` and the rule's own
// termination.
//
// end must be non-zero and strictly after start; a zero or
// inverted window is a caller error (wrapped ErrUnboundedExpansion)
// rather than a silent empty result — an unbounded window would be
// an infinite expansion request.
//
// The window bounds the result, not the search: a rule that never
// yields never reaches `end`, so the evaluator's iteration bound
// stops it and Between reports a wrapped ErrIterationCap.
//
// Occurrences before dtstart never exist, so a window entirely
// before dtstart returns an empty slice.
func Between(rule Rule, dtstart, start, end time.Time) ([]time.Time, error) {
	if end.IsZero() {
		return nil, fmt.Errorf("rrule: Between: end must be non-zero (an open-ended window is an infinite expansion; use All): %w", ErrUnboundedExpansion)
	}
	if !end.After(start) {
		return nil, fmt.Errorf("rrule: Between: end %s must be after start %s: %w", end, start, ErrUnboundedExpansion)
	}
	if err := checkExpandable(rule); err != nil {
		return nil, err
	}

	out := []time.Time{}
	capped := walk(rule, dtstart, func(occ time.Time) bool {
		if !occ.Before(end) {
			return false
		}
		if !occ.Before(start) {
			out = append(out, occ)
		}
		return true
	})
	if capped {
		return nil, iterationCapError("Between", rule)
	}
	return out, nil
}

// walk is the shared generator behind All, Occurrences, Between and
// Set. It yields every occurrence of rule from dtstart in
// chronological order and reports how the walk ended: capped is
// true when MaxIterations consecutive FREQ periods produced no
// occurrence and the search was abandoned; false when the rule
// terminated (UNTIL, COUNT, or no further period) or yield stopped
// it.
//
// The bound is a starvation guard, not a total-occurrence limit:
// the empty-period counter resets whenever a period yields, so a
// rule that fires regularly runs indefinitely (bounded only by the
// caller's limit, window, or break) while one whose BY-* filters
// can never match still stops. It shares MaxIterations with
// NextOccurrence so the package documents one bound.
//
// An invalid rule (FREQ unset, INTERVAL < 1) yields nothing and is
// not a cap hit; callers that need it reported run checkExpandable
// first.
func walk(rule Rule, dtstart time.Time, yield func(time.Time) bool) (capped bool) {
	if checkExpandable(rule) != nil {
		return false
	}
	emitted := 0
	empty := 0
	current := dtstart

	for empty < MaxIterations {
		occs := periodOccurrences(rule, current, dtstart)
		produced := false
		for _, occ := range occs {
			if occ.Before(dtstart) {
				continue
			}
			// UNTIL is inclusive per RFC 5545 §3.3.10.
			if !rule.Until.IsZero() && occ.After(rule.Until) {
				return false
			}
			produced = true
			emitted++
			if !yield(occ) {
				return false
			}
			if rule.Count > 0 && emitted >= rule.Count {
				return false
			}
		}
		if produced {
			empty = 0
		} else {
			empty++
		}
		next, ok := advance(rule, current)
		if !ok {
			return false
		}
		current = next
	}
	return true
}

// iterationCapError wraps ErrIterationCap with the entry point and
// rule context, mirroring NextOccurrence's own message shape.
func iterationCapError(op string, rule Rule) error {
	return fmt.Errorf("rrule: %s: no occurrence within %d consecutive %s periods: %w",
		op, MaxIterations, rule.Freq, ErrIterationCap)
}

// checkExpandable rejects Rule values the evaluator cannot walk.
// ParseRRule guarantees both conditions hold; a hand-built Rule
// literal may not.
func checkExpandable(rule Rule) error {
	if rule.Freq == FreqInvalid {
		return fmt.Errorf("rrule: FREQ is required: %w", ErrUnsupportedRRule)
	}
	if rule.Interval < 1 {
		return fmt.Errorf("rrule: INTERVAL must be >= 1, got %d: %w", rule.Interval, ErrUnsupportedRRule)
	}
	return nil
}

// periodOccurrences expands one FREQ period into its concrete
// occurrences, sorted chronologically and with BYSETPOS applied.
//
// It is the shared kernel behind walk (and therefore All,
// Occurrences, Between, and Set): the per-period expansion plus
// ordering plus positional filtering that RFC 5545 §3.3.10
// specifies, factored out so the iteration strategy above it can
// vary.
func periodOccurrences(rule Rule, base, dtstart time.Time) []time.Time {
	occs := expand(rule, base, dtstart)
	slices.SortFunc(occs, func(a, b time.Time) int { return a.Compare(b) })
	if len(rule.BySetPos) > 0 {
		occs = applyBySetPos(occs, rule.BySetPos)
	}
	return occs
}
